File: src/sort/insertion-sort.ts
export class InsertionSort {

  /**
   * The implementation of insertion sort
   * @param arr The array to be sorted
   */
  public static insertionSort (arr: number[]) {
    for (let i = 0; i < arr.length; i++) {
      // Find the appropriate insertion position for the arr[i] element
      for (let j = i; j > 0; j--) {
        if (arr[j] < arr[j - 1]) {
          const temp   = arr[j]
          arr  [j]     = arr[j - 1]
          arr  [j - 1] = temp
        } else {
          break
        }
      }
    }
  }

  /**
   * The optimized implementation of insertion sort
   * @param arr The array to be sorted
   */
  public static insertionSortOptimization (arr: number[]) {
    for (let i = 1; i < arr.length; i++) {
      // Find the appropriate insertion position for the arr[i] element
      let   j            = 0
      const currentValue = arr[i]
      for (j = i; j > 0; j --) {
        if (currentValue < arr[j - 1]) {
          arr[j] = arr[j - 1]
        } else {
          break
        }
      }
      arr[j] = currentValue
    }
  }
}

File: java/src/main/java/club/linyimin/algorithm/sort/insertion/InsertionSort.java
package club.linyimin.algorithm.sort.insertion;

import club.linyimin.algorithm.sort.Sort;

/**
 * @author yiminlin
 * @date 2021/12/19 7:39 pm
 * @description insertion sort implementation
 **/
public class InsertionSort implements Sort {

    @Override
    public int[] sort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            for (int j = i; j > 0; j--) {
                if (arr[j] >= arr[j-1]) {
                    break;
                }
                int temp = arr[j];
                arr[j - 1] = temp;
                arr[j] = arr[j - 1];
            }
        }
        return arr;
    }
}
